agent/agentssh/agentssh.go

```diff
@@ -0,0 +1,571 @@
+package agentssh
+
+import (
+	"bufio"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/exec"
+	"os/user"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gliderlabs/ssh"
+	"github.com/pkg/sftp"
+	gossh "golang.org/x/crypto/ssh"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/agent/usershell"
+	"github.com/coder/coder/codersdk/agentsdk"
+	"github.com/coder/coder/pty"
+)
+
+const (
+	// MagicSessionErrorCode indicates that something went wrong with the session, rather than the
+	// command just returning a nonzero exit code, and is chosen as an arbitrary, high number
+	// unlikely to shadow other exit codes, which are typically 1, 2, 3, etc.
+	MagicSessionErrorCode = 229
+
+	// MagicSessionTypeEnvironmentVariable is used to track the purpose behind an SSH connection.
+	// This is stripped from any commands being executed, and is counted towards connection stats.
+	MagicSessionTypeEnvironmentVariable = "CODER_SSH_SESSION_TYPE"
+	// MagicSessionTypeVSCode is set in the SSH config by the VS Code extension to identify itself.
+	MagicSessionTypeVSCode = "vscode"
+	// MagicSessionTypeJetBrains is set in the SSH config by the JetBrains extension to identify itself.
+	MagicSessionTypeJetBrains = "jetbrains"
+)
+
+type Server struct {
+	serveWg sync.WaitGroup
+	logger  slog.Logger
+
+	srv *ssh.Server
+
+	Env        map[string]string
+	AgentToken func() string
+
+	manifest atomic.Pointer[agentsdk.Manifest]
```

Comment:
Since you already have an atomic.Pointer for manifest in the agent, I think you should change both to be `*atomic.Pointer[agentsdk.Manifest]` and use the same underlying pointer for both. This is how we share things with coderd and provisionerdserver for example.

Reply:
you could avoid SetManifest in this case, as the agent would just swap into it's own pointer

Reply:
Great suggestion! I suppose making it a pointer in `agentssh` is sufficient, and then passing via `sshSrv.Manifest = &agent.manifest`